### PR TITLE
[sprites needed]Adds the rebreather tank, a high-tech internals device.

### PR DESCRIPTION
**Pull request by Tacolizard**
:cl: NanoTrasen EVA Club
add: Added the rebreather tank, an internals tank that never runs out of air! Well, as long as you keep it charged.
add: Gave the CE a rebreather
add: Gave Atmos Techs rebreathers
tweak: The spacesuit crate was renamed the EVA crate
add: Added two rebreathers to the EVA crate
/:cl:

Inspired by [these](https://en.wikipedia.org/wiki/Rebreather), the rebreather tank is an internals tank that uses electrical charge to 'filter' air. It has two settings, human and plasmaman. The human setting fills the tank with o2 and scrubs plasma, the plasmaman setting does vice versa. The tank can be recharged in a standard weapons recharger.

Because the latest meme seems to be hating on people who give things to RnD, I have opted to make this item non-researchable. It is only obtainable by being the CE or ordering through cargo.

TODO:
[] Item icon
[] Inhand icons
[] on-mob icons

**Please someone make sprites or I'll have to do them. And we all know how [that](https://camo.githubusercontent.com/56c77721cb783c155c7bbc584a2b50b78c2b5940/68747470733a2f2f692e6779617a6f2e636f6d2f61336565616531666636613732383435363464356331636233353535313365622e706e67) can turn out.**

**now rebalanced to be even harder to use for spacebases**

:taco:

### CODE REVIEW

**Comment by kevinz000** [comment]
im deciding whether to downvote

**Comment by Tacolizard** [comment]
why

**Comment by ShadowDeath6** [comment]
dude it's high tech give it to R&D

alternatively- more CE powercreep

**Comment by Tacolizard** [comment]
@ShadowDeath6 I am firmly in the position that any CE powercreep is good powercreep.

Also, I really want to give it to RnD, but seeing what happened to the last few PRs doing that I think it would be easier not to.

File: code/game/machinery/recharger.dm [hunk 1]
```diff
@@ -124,6 +124,13 @@
 					use_power(200 * recharge_coeff)
 					using_power = 1
 
+		if(istype(charging, /obj/item/weapon/tank/internals/rebreather))
+			var/obj/item/weapon/tank/internals/rebreather/B = charging
+			if(B.tcell)
+				if(B.tcell.give(B.tcell.chargerate * recharge_coeff))
+					use_power(200 * recharge_coeff)
+					using_power = 1
```
Comment (Expletive) [inline code/game/machinery/recharger.dm]: It's nice seeing a new battery powered device. But there's gonna be merge conflicts with the inducer.  I just got done refactoring it to use get_cell(), though, so this whole block won't be necessary if rebreathers are in allowed_devices.

File: code/modules/jobs/job_types/engineering.dm [hunk 2]
```diff
@@ -145,6 +146,7 @@ Atmospheric Technician
 	box = /obj/item/weapon/storage/box/engineer
 	pda_slot = slot_l_store
 	backpack_contents = list(/obj/item/device/modular_computer/tablet/preset/advanced=1)
+	l_hand = /obj/item/weapon/tank/internals/rebreather
```
Comment (Expletive) [inline code/modules/jobs/job_types/engineering.dm]: Why putting it in their hands and not their backpacks?

**Comment by Bawhoppen** [comment]
no more R&D items don't cave to the inevitable pressure

File: code/modules/jobs/job_types/engineering.dm (hunk 2 above)
Comment (Tacolizard) [inline code/modules/jobs/job_types/engineering.dm]: 1. When I wrote this I assumed that they don't fit in backpacks
2. It's a very useful item, so hopefully people will try to use it if it's in their hand

**Comment by XDTM** [comment]
CE is gonna have more toys than the other heads combined soon

**Comment by kevinz000** [comment]
@Bawhoppen im an exception to the rule 
@XDTM i want more stuff as rd/hos ;-;

**Comment by XDTM** [comment]
Also i know the stigma against rnd, but this would make for a fancy cybernetix implant

**Comment by Expletive** [comment]
1) Have you done any tests to see how this compares to regular oxygen tanks? 
A) How long does a fully charged rebreather last compared to a standard tank pressurized at 607kpa? 
B) At 1013kpa? 
C) At 3039kpa?

2) Have you investigated any of the bugs that might arise from attaching it to a tank transfer valve?
3) What about using it to recharge a jetpack via a tank transfer valve? 
4) What about if you put the rebreather into a canister or air pump and pressurize it?
5) Have you considered making the battery replaceable?

**Comment by ShadowDeath6** [comment]
I didn't think you could use internals tanks on TTVs

**Comment by Tacolizard** [comment]
@Expletive The method it uses to refill itself is super simple, there shouldn't be any issues with TTVs or jetpacks. It is extremely similar to a normal tank. As for canister pressurization, that works and is part of the design.

**Comment by FantasticFwoosh** [comment]
I am also in the same boat about just shoving a airtank in a recharger as @kevinz000 , it seems very very cheap, especially when the charger is upgraded in order to cheese on atmos.

You know portable air cons can actually exhume air right? You could abuse the tanks to either refill a empty canister full of air, and it'd be undoubtedly overpowered in areas with unlimited power (white ship, mining pods etc) with a preset or constructed recharger board within.

> Portable air cons can also suck all the air out of a room, so for every 100 units of ambient pressure from sucking away the internals of a room, a exploit is to just to set the aircon to 300 and let it run sucking up the atmos around it to turn into oxygen, its game breakingly good.

In any such case this has a good chance to be merged miners should probably be supplied with them, infact if you made all the extended capacity emergency air tanks the same size as regular one and just allowed them to recharge, you could replace them entirely.

**Comment by Tacolizard** [comment]
@FantasticFwoosh if you read my code, you would see that this does not in any way 'suck 100 units of ambient pressure'. Please familiarize yourself with the difference between moles, kPa, and cell charge ratings.

Are you aware that atmos has a virtually unlimited supply of gas roundstart? There is no point in using a rebreather to fill an air can. It would also be extremely slow, and only net you around +34 kPa of however much gas you started with, as the rebreather requires its pressure to be below 100 before it starts refilling, so you can't generate infinite gas.

**Comment by FantasticFwoosh** [comment]
In that case, empty with a portable air con & slap inside a upgraded recharger, repeat. Its spontaneous & valuable air that you could use from a mining pod to quickly make places breathable is where i see the exploit.

**Comment by Tacolizard** [comment]
or... you could bring an oxygen canister and set it to output at 101kPa to quickly make places breathable, much more simply and efficiently.

Also you have not addressed the fact that atmos has a virtually unlimited supply of gas roundstart, so there's very little point in 'exploiting' things like rebreathers.

**Comment by Tacolizard** [comment]
>valuable air

**Comment by FantasticFwoosh** [comment]
Im not talking about the station per say. If you could quickly re-oxygenate any place on or off on the station it will probably be used for space exploration or to build bases such as clock cult, offstation but still on the edges of the Z level

> Throw a maxcap at the air canisters in maint & in atmos, suddenly choke to death because your 4+ hours supply of air just got wiped out.

**Comment by Tacolizard** [comment]
you do know you can't 'oxygenate space' right?

**Comment by Tacolizard** [comment]
even if you could, did you even think for a second about how long it would take, 100kPa at a time, to fill THE ENTIRETY OF SPACE with air?

**Comment by FantasticFwoosh** [comment]
Use your imagination, i meant enclosed room in space probably through a clock cult/blood cult portal connected by teleportation & a mining pod for infinite recharging and a looted air dispenser to make a breathable area of operations.

> Its already possible but this makes it trivially easy

**Comment by Tacolizard** [comment]
TRIVIALLY EASY?

DO YOU KNOW HOW LITTLE 100 FUCKING KPA IS WHEN YOU JUST LET IT LOOSE IN A VACUUM? WHY ON EARTH OR ANY OTHER PLANET IN ANY DIMENSION EVER WOULD YOU WANT TO AND GO THROUGH THE TROUBLE OF DOING THAT?

# WHAT THE FUCK

HOW THE FUCK IS

1. OBTAINING AND FULLY UPGRADING A WEAPONS DOCK
2. STEALING OR CONVINCING CARGO TO ORDER A REBREATHER
3. STEALING A GAS PUMP FROM ATMOS, THE SAME PLACE YOU CAN GET GAS CANS
4. USING THE REBREATHER TO FILL A COLD VACUUM 2.9 MOLES AT A TIME

EASIER THAN
1. GRABBING A FUCKING GAS CAN
2. OPENING IT
3. ????
4. PROFIT

**Comment by FantasticFwoosh** [comment]
Because sempai you can just use a rapid pipe dispenser, make a canister port and put the air con on it within your pod next to your recharger wrenched in. Some simple piping then leads to a air canister built out of raw metal outside also wrenched into that mini pipe network.

Then simply leave the air con on constantly to the full potential contents of the air tank to add to the pipe network, this will fill the air tank FOR NOTHING. And release the airtank at 101 infinitely forever whenever you need it because its connected to a infinite supply of air and spend time recharging the rebreather or multiple rebreathers to refill the airtank in time back to full.

> On box there's a recharger in the security checkpoint outside arrivals, smash it with a toolbox until it deconstructs (or decon it, most rechargers are not out of sight) & grab the board, deconstruct the frame if you need to

**Comment by kevinz000** [comment]
stop whining about infinite air i'm going to add actual infinite air that can be used to easily repressurize a space base/mining base soon his is just slightly more portable.
and stop bitching in all caps that's my job
you two sound silly tbh.

**Comment by optimumtact** [comment]
This is extremely overpowered it's a 👎 from me

**Comment by Tacolizard** [comment]
@optimumtact atmos is a source of infinite oxygen. this is infinite oxygen with some limits. how is this OP?

**Comment by SgtxRootbeer** [comment]
Wait, does this work with hardsuit jetpacks?

**Comment by Tacolizard** [comment]
@SgtxRootbeer I don't see why not

**Comment by kevinz000** [comment]
:-1: ^

**Comment by XDTM** [comment]
Never argue with fwoosh

**Comment by Tacolizard** [comment]
>In that case, empty with a portable air con & slap inside a upgraded recharger, repeat. Its spontaneous & valuable air that you could use from a mining pod to quickly make places breathable is where i see the exploit.

+ if(air_contents)

I read this as applying to any air contents inside the tank, but i see you changed it on the second to final commit. The first one would have been exploitative.

Do you know what that even means?

if(air_contents) is just checking if the canister has any gases in it

**Comment by FantasticFwoosh** [comment]
Booted up a test server, 3 

Tanks need to be removed & recharged every 40 seconds, giving out 700-800kpa each, which means in 5 cycles with 3 rebreathers you will have completely filled to a equal level of a oxygen tank. 

Nitrogen is separated but essentially this is making free pure oxygen out of energy slowly but to as many re-breathers that are hooked up

**Comment by SgtxRootbeer** [comment]
Then the solution is simple, give the rebreather a pressure of 39kpa, instead of 100, to make it unfeasable, and change the output to pure o2.
Alternatively, don't care because​ blood cults have hardsuits already, and if they can get slime cells, they can also get o2 canisters.
And clock cults never make space base, they either use the ai core, or some room in maint.

File: code/game/machinery/recharger.dm [hunk 3]
```diff
@@ -8,7 +8,7 @@
 	idle_power_usage = 4
 	active_power_usage = 250
 	var/obj/item/charging = null
-	var/list/allowed_devices = list(/obj/item/weapon/gun/energy,/obj/item/weapon/melee/baton,/obj/item/ammo_box/magazine/recharge,/obj/item/device/modular_computer)
+	var/list/allowed_devices = list(/obj/item/weapon/gun/energy,/obj/item/weapon/melee/baton,/obj/item/ammo_box/magazine/recharge,/obj/item/device/modular_computer, /obj/item/weapon/tank/internals/rebreather)
```
Comment (Cyberboss) [inline code/game/machinery/recharger.dm]: static list

File: code/game/objects/items/weapons/tanks/tank_types.dm [hunk 4]
```diff
@@ -37,6 +38,50 @@
...
+//Rebreather
+/obj/item/weapon/tank/internals/rebreather
+	name = "rebreather tank"
+	desc = "An advanced improvement on the internals tank that uses electricity to recycle ambient waste gases into breathable ones. AltClick it to turn it on or switch target species."
+	icon_state = "oxygen"
+	distribute_pressure = TANK_DEFAULT_RELEASE_PRESSURE
+	force = 10
+	dog_fashion = /datum/dog_fashion/back
+	var/species = "off"
+	var/obj/item/weapon/stock_parts/cell/high/tcell = new /obj/item/weapon/stock_parts/cell/high //maxcharge is 10000
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/tanks/tank_types.dm]: just `= new`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by FantasticFwoosh** [comment]
@SgtxRootbeer the AI core is a premade space base of sorts detached from the station on box & meta.

But the pure O2 output wont do anything except make the piping exploitation even easier, when the fans are running they consume 32kpa and then the cell refreshes it again. It actually wouldn't change the pace in which the canister in that setup as the example i've already shown you because its constantly regenerating until its 

> The fact it regenerates from energy and replaces it is the sole point of the exploit.

The gas divider is just there to clean the atmos if the pumps sucked up anything else by accident, and you could probably offshoot it to a scrubber in space.

**Comment by Tacolizard** [comment]
@SgtxRootbeer good suggestion, the rebreather now only refills below 39kPa, making fwoosh's spacebase 'exploit' even more unfeasible. Also, it always only outputs o2 on human mode.

File: code/game/machinery/recharger.dm (hunk 3 above)
Comment (Tacolizard) [inline code/game/machinery/recharger.dm]: What's a static list

**Comment by Tacolizard** [comment]
@FantasticFwoosh for like the fifth time, you are not 'harvesting' oxygen. It doesn't 'scrub' anything.

**Comment by FantasticFwoosh** [comment]
No, thats what the piping system is doing, just cleaning out anything extra that might have been in the pipes. Really at this point where i've actually brought evidence of a exploit to your face and you still wriggle around into denying it. Pod has infinite energy and the air pumps suck up 30+ kpa every tick whilst regenerating normally.

> You still haven't fixed the exploit.

As rude to title a commit after me too. Ew.

**Comment by kevinz000** [comment]
@XDTM yeah just pretend he didn't say anything unless it's constructive (oh wait..)

**Comment by PKPenguin321** [comment]
Sounds cool

**Comment by lzimann** [comment]
Ping whenever you have the sprites I guess.

**Comment by JJRcop** [comment]
revive this

**Pull request closed without merging**